module/VuFind/src/VuFind/Recommend/Databases.php

```diff
@@ -0,0 +1,275 @@
+    {
+        $settings = explode(':', $settings);
+        $this->limit
+            = (isset($settings[0]) && is_numeric($settings[0]) && $settings[0] > 0)
+            ? intval($settings[0]) : 5;
+        $databasesConfigFile = $settings[1] ?? 'EDS';
+
+        $databasesConfig = $this->configManager->get($databasesConfigFile);
+        $configUrls = isset($databasesConfig->Databases->url)
+            ? $databasesConfig->Databases->url->toArray() : [];
+        $this->configFileDatabases = array_map(function ($url) {
+            return ['url' => $url];
+        }, $configUrls);
+
+        $this->resultFacet = isset($databasesConfig->Databases->resultFacet)
+            ? $databasesConfig->Databases->resultFacet->toArray() : [];
+        $this->resultFacetNameKey = $databasesConfig->Databases->resultFacetNameKey ?? 'value';
+
+        $this->useLibGuides = $databasesConfig->Databases->useLibGuides ?? false;
+        if ($this->useLibGuides) {
+            // Cache the data related to profiles for up to 10 minutes:
+            $libGuidesApiConfig = $this->configManager->get('LibGuidesAPI');
+            $this->cacheLifetime = intval($libGuidesApiConfig->GetAZ->cache_lifetime ?? 600);
+        }
+    }
+
+    /**
+     * Called before the Search Results object performs its main search
+     * (specifically, in response to \VuFind\Search\SearchRunner::EVENT_CONFIGURED).
+     * This method is responsible for setting search parameters needed by the
+     * recommendation module and for reading any existing search parameters that may
+     * be needed.
+     *
+     * @param \VuFind\Search\Base\Params $params  Search parameter object
+     * @param \Laminas\Stdlib\Parameters $request Parameter object representing user
+     * request.
+     *
+     * @return void
+     *
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function init($params, $request)
+    {
+        // No action needed.
+    }
+
+    /**
+     * Called after the Search Results object has performed its main search. This
+     * may be used to extract necessary information from the Search Results object
+     * or to perform completely unrelated processing.
+     *
+     * @param \VuFind\Search\Base\Results $results Search results object
+     *
+     * @return void
+     */
+    public function process($results)
+    {
+        $this->results = $results;
+    }
+
+    /**
+     * Get terms related to the query.
+     *
+     * @return array
+     */
+    public function getResults()
+    {
+        if (count($this->resultFacet) < 1) {
+            throw new \Exception('At least one facet key required.');
+        }
+
+        $resultDatabasesTopFacet = array_shift($this->resultFacet);
+        $resultDatabases = $this->results->getFacetList([$resultDatabasesTopFacet => null])[$resultDatabasesTopFacet];
+        while (count($this->resultFacet) && $resultDatabases) {
+            // Throw exception if the key is not found
```

Comment:
You have some comments about throwing exceptions, but I don't see any exceptions being thrown (unless you're referring to the native PHP errors caused by accessing undefined indexes).

Reply:
Yes I was referring to the native exceptions -- I thought they were reasonable to let bubble up here; I don't think any defaults / fallbacks are helpful.  If it's better I can throw explicit exceptions.

Reply:
My main concern is that the behavior may not be predictable depending on the error_reporting settings in php.ini. It's possible that this will throw a fatal error when you're in development mode (with zero tolerance for PHP notices) but this will silently ignore the error in production mode (when the settings are generally more error-tolerant). I don't think that's the desired behavior! But maybe I'm misremembering what counts as a notice and what doesn't, so feel free to correct me if you've already tested this and I'm wrong. :-)

Reply:
Huh no I didn't think of that; have been testing in production mode and looking at the logs.  I'm honestly not sure the right approach, since in my experience VuFind seems to use defaults to avoid exceptions, and I don't think that's right here.  Not sure how best to handle it.

Reply:
I guess the question is, what's the desired behavior and intended goal here? Is the idea that if the user configures an invalid path to facet data, it causes a fatal error? If so, is that really desirable? I'm not sure that it would actually make this any easier to troubleshoot. Might it be better if we use null coalescing to treat missing data as empty so that it will be ignored without exceptions, and perhaps add some debug logging in cases where an expected configured key is undefined?

Reply:
Ok that's a good way to look at it -- I think it's better to log an error than break everything.  Presuming anyone is paying attention to the log output. :)  I am not using null coalescing here though because there is no default handling of these keys not being found that should not cause at least a log message; in theory I could coalesce to null, check for null, and log the error, but using try/catch makes it clearer IMHO.